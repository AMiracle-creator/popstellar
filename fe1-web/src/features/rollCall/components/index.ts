export { default as RollCallListItem, RollCallEventType } from './RollCallListItem';
