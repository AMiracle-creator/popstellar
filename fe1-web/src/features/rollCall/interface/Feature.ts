import { Hash, PublicKey } from 'core/objects';

export namespace RollCallFeature {
  export interface Lao {
    id: Hash;

    witnesses: PublicKey[];
  }

  export interface Event {
    id: Hash;
    name: string;
  }
}
