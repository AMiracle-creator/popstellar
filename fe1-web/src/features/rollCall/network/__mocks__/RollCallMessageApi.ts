/**
 * Mocks functions to send roll call messages.
 */
export const requestCloseRollCall = jest.fn(() => Promise.resolve());
