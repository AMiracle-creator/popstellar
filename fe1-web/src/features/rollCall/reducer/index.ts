export { default as rollCallReducer } from './RollCallReducer';
export * from './RollCallReducer';
