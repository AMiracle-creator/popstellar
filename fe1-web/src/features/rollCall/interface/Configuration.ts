import React from 'react';
import { AnyAction, Reducer } from 'redux';

import { MessageRegistry } from 'core/network/jsonrpc/messages';
import { Hash, PopToken } from 'core/objects';
import FeatureInterface from 'core/objects/FeatureInterface';

import { RollCallReducerState, ROLLCALL_REDUCER_PATH } from '../reducer';
import { RollCallFeature } from './Feature';

export const ROLLCALL_FEATURE_IDENTIFIER = 'rollCall';

export interface RollCallConfiguration {
  // objects
  messageRegistry: MessageRegistry;

  /* LAO related functions */

  /**
   * Gets the lao associated to the given id. Should be used outside react components
   */
  getLaoById: (id: string) => RollCallFeature.Lao | undefined;

  /**
   * Returns the currently active lao id. Should be used inside react components
   * @returns The current lao id
   */
  useCurrentLaoId: () => Hash | undefined;

  /**
   * An action cretor that sets the last roll call for a given lao
   */
  setLaoLastRollCall: (
    laoId: Hash | string,
    rollCallId: Hash | string,
    hasToken: boolean,
  ) => AnyAction;

  /* Event related functions */

  /**
   * Creates a redux action for adding an event to the event store
   * @param laoId - The lao id where to add the event
   * @param eventState - The event to add to the store
   * @returns A redux action causing the state change
   */
  addEvent: (laoId: string | Hash, eventState: RollCallFeature.EventState) => AnyAction;

  /**
   * Creates a redux action for update the stored event state
   * @param laoId - The lao id where to update the event
   * @param eventState - The update event state
   */
  updateEvent: (laoId: string | Hash, eventState: RollCallFeature.EventState) => AnyAction;

  /**
   * Given the redux state and an event id, this function looks in the active
   * lao for an event with a matching id, creates an instance of the corresponding type
   * and returns it
   * @param id - The id of the event
   * @returns The event or undefined if none was found
   */
  getEventById: (id: Hash) => RollCallFeature.Event | undefined;

  /**
   * Creates a selector to return a specific event for given lao and event ids
   * @param laoId The id of the lao to select the event from
   * @param eventId The id of the event to return
   * @returns The selector
   */
  makeEventSelector: (
    laoId: Hash | string,
    eventId: Hash | string,
  ) => (state: unknown) => RollCallFeature.Event | undefined;

  /**
   * Deterministically generates a pop token from given lao and rollCall ids
   * @param laoId The lao id to generate a token for
   * @param rollCallId The rollCall id to generate a token for
   * @returns The generated pop token
   */
  generateToken: (laoId: Hash, rollCallId: Hash | undefined) => Promise<PopToken>;

  /**
   * Checks whether a seed is present in the wallet store
   */
  hasSeed: () => boolean;
}

/**
 * The type of the context that is provided to react rollcall components
 */
export type RollCallReactContext = Pick<
  RollCallConfiguration,
  'useCurrentLaoId' | 'makeEventSelector' | 'generateToken' | 'hasSeed'
>;

/**
 * The interface the rollcall feature exposes
 */
export interface RollCallInterface extends FeatureInterface {
  screens: {
    CreateRollCall: React.ComponentType<any>;
    RollCallOpened: React.ComponentType<any>;
  };

  eventTypeComponents: {
    isOfType: (event: unknown) => boolean;
    Component: React.ComponentType<{ event: unknown; isOrganizer: boolean | null | undefined }>;
  }[];

  context: RollCallReactContext;

  reducers: {
    [ROLLCALL_REDUCER_PATH]: Reducer<RollCallReducerState>;
  };
}
