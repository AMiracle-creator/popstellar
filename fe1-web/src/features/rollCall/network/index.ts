import { ActionType, ObjectType } from 'core/network/jsonrpc/messages';
import { Hash } from 'core/objects';
import { dispatch, getStore } from 'core/redux';

import { RollCallConfiguration } from '../interface';
import { RollCall } from '../objects';
import { addRollCall, getRollCallById, updateRollCall } from '../reducer';
import { CloseRollCall, CreateRollCall, OpenRollCall, ReopenRollCall } from './messages';
import {
  handleRollCallCloseMessage,
  handleRollCallCreateMessage,
  handleRollCallOpenMessage,
  handleRollCallReopenMessage,
} from './RollCallHandler';

export * from './RollCallMessageApi';

/**
 * Configures the network callbacks in a MessageRegistry.
 *
 * @param configuration - The configuration object for the rollcall feature
 */
export const configureNetwork = (configuration: RollCallConfiguration) => {
  // getRollCallById bound to the global state
  const boundGetRollCallById = (rollCallId: Hash | string) =>
    getRollCallById(rollCallId, getStore().getState());

  const addRollCallEvent = (laoId: Hash | string, rollCall: RollCall) => {
    const rollCallState = rollCall.toState();

    dispatch(
      configuration.addEvent(laoId, {
        eventType: RollCall.EVENT_TYPE,
        id: rollCallState.id,
        idAlias: rollCallState.idAlias,
        start: rollCall.start.valueOf(),
        end: rollCall.end.valueOf(),
      }),
    );
    dispatch(addRollCall(rollCallState));
  };

  const updateRollCallEvent = (rollCall: RollCall) => {
    const rollCallState = rollCall.toState();

    dispatch(
      configuration.updateEvent({
        eventType: RollCall.EVENT_TYPE,
        id: rollCallState.id,
        idAlias: rollCallState.idAlias,
        start: rollCall.start.valueOf(),
        end: rollCall.end.valueOf(),
      }),
    );
    dispatch(updateRollCall(rollCallState));
  };

  configuration.messageRegistry.add(
    ObjectType.ROLL_CALL,
    ActionType.CREATE,
    handleRollCallCreateMessage(addRollCallEvent),
    CreateRollCall.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ROLL_CALL,
    ActionType.OPEN,
    handleRollCallOpenMessage(boundGetRollCallById, updateRollCallEvent),
    OpenRollCall.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ROLL_CALL,
    ActionType.CLOSE,
    handleRollCallCloseMessage(
      boundGetRollCallById,
      updateRollCallEvent,
      configuration.generateToken,
      configuration.setLaoLastRollCall,
    ),
    CloseRollCall.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ROLL_CALL,
    ActionType.REOPEN,
    handleRollCallReopenMessage(boundGetRollCallById, updateRollCallEvent),
    ReopenRollCall.fromJson,
  );
};
