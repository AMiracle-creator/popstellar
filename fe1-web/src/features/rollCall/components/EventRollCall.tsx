import { useNavigation } from '@react-navigation/native';
import PropTypes from 'prop-types';
import React, { useEffect, useState, FunctionComponent, useMemo } from 'react';
import { Text } from 'react-native';
import { useToast } from 'react-native-toast-notifications';
import { useSelector } from 'react-redux';

import { QRCode, WideButtonView } from 'core/components';
import { FOUR_SECONDS } from 'resources/const';
import STRINGS from 'resources/strings';

import { RollCallHooks } from '../hooks';
import { requestOpenRollCall, requestReopenRollCall } from '../network';
import { RollCall, RollCallStatus } from '../objects';
import { makeRollCallSelector } from '../reducer';

/**
 * Component used to display a RollCall event in the LAO event list
 */
const EventRollCall = (props: IPropTypes) => {
  const { eventId: rollCallId, isOrganizer } = props;

  const selectRollCall = useMemo(() => makeRollCallSelector(rollCallId), [rollCallId]);
  const rollCall = useSelector(selectRollCall);

  if (!rollCall) {
    throw new Error(`Could not find a roll call with id ${rollCallId}`);
  }

  // FIXME: use a more specific navigation
  const navigation = useNavigation<any>();
  const toast = useToast();

  const laoId = RollCallHooks.useCurrentLaoId();
  const generateToken = RollCallHooks.useGenerateToken();
  const hasSeed = RollCallHooks.useHasSeed();

  if (!laoId) {
    throw new Error('no LAO is currently active');
  }
  const [popToken, setPopToken] = useState('');
  const [hasWalletBeenInitialized, setHasWalletBeenInitialized] = useState(hasSeed());

  // re-check if wallet has been initialized after focus events
  useEffect(() => {
    // Return the function to unsubscribe from the event so it gets removed on unmount
    return navigation.addListener('focus', () => {
      setHasWalletBeenInitialized(hasSeed());
    });
  }, [navigation, hasSeed]);

  // Once the roll call is opened the first time, idAlias is defined, and needed for closing/reopening the roll call
  const eventHasBeenOpened = rollCall.idAlias !== undefined;

  useEffect(() => {
    if (!hasWalletBeenInitialized || !laoId || !rollCall?.id) {
      return;
    }

    // Here we get the pop-token to display in the QR code
    generateToken(laoId, rollCall.id)
      .then((token) => setPopToken(token.publicKey.valueOf()))
      .catch((err) => console.error(`Could not generate token: ${err}`));
  }, [hasWalletBeenInitialized, generateToken, laoId, rollCall]);

  if (!rollCall) {
    console.debug('Error in Roll Call display: Roll Call doesnt exist in store');
    return null;
  }

  const makeToastErr = (error: string) => {
    toast.show(error, {
      type: 'danger',
      placement: 'bottom',
      duration: FOUR_SECONDS,
    });
  };

  const onOpenRollCall = () => {
    requestOpenRollCall(laoId, rollCall.id).catch((e) => {
      makeToastErr('Unable to send roll call open request');
      console.debug('Unable to send Roll call open request', e);
    });
  };

  const onReopenRollCall = () => {
    if (eventHasBeenOpened) {
      requestReopenRollCall(laoId, rollCall.idAlias).catch((e) => {
        makeToastErr('Unable to send Roll call re-open request');
        console.debug('Unable to send Roll call re-open request', e);
      });
    } else {
      makeToastErr('Unable to send roll call re-open request, the event does not have an idAlias');
      console.debug('Unable to send roll call re-open request, the event does not have an idAlias');
    }
  };

  const onScanAttendees = () => {
    if (eventHasBeenOpened) {
      navigation.navigate(STRINGS.roll_call_open, {
        rollCallID: rollCall.id.toString(),
      });
    } else {
      makeToastErr('Unable to scan attendees, the event does not have an idAlias');
      console.debug('Unable to scan attendees, the event does not have an idAlias');
    }
  };

  const getRollCallDisplay = (status: RollCallStatus) => {
    switch (status) {
      case RollCallStatus.CREATED:
        return (
          <>
            <Text>Not Open yet</Text>
            <Text>Be sure to have set up your Wallet</Text>
            {isOrganizer && (
              <WideButtonView title={STRINGS.roll_call_open} onPress={() => onOpenRollCall()} />
            )}
          </>
        );
      case RollCallStatus.REOPENED:
      case RollCallStatus.OPENED:
        return (
          <>
            {!isOrganizer && (
              <>
                <Text>Let the organizer scan your Pop token</Text>
                <QRCode visibility value={popToken} />
              </>
            )}
            {isOrganizer && (
              <WideButtonView
                title={STRINGS.roll_call_scan_attendees}
                onPress={() => onScanAttendees()}
              />
            )}
          </>
        );
      case RollCallStatus.CLOSED:
        return (
          <>
            <Text>Closed</Text>
            <Text>Attendees are:</Text>
            {rollCall.attendees?.map((attendee) => (
              <Text key={attendee.valueOf()}>{attendee}</Text>
            ))}
            {isOrganizer && (
              <WideButtonView title={STRINGS.roll_call_reopen} onPress={() => onReopenRollCall()} />
            )}
          </>
        );
      default:
        console.warn('Roll Call Status was undefined in EventRollCall');
        return null;
    }
  };

  return (
    <>
      <Text>Roll Call</Text>
      {getRollCallDisplay(rollCall.status)}
    </>
  );
};

const propTypes = {
  eventId: PropTypes.string.isRequired,
  isOrganizer: PropTypes.bool,
};
EventRollCall.propTypes = propTypes;

EventRollCall.defaultProps = {
  isOrganizer: false,
};

type IPropTypes = PropTypes.InferProps<typeof propTypes>;

export default EventRollCall;

export const RollCallEventType = {
  eventType: RollCall.EVENT_TYPE,
  navigationNames: {
    createEvent: STRINGS.organizer_navigation_creation_roll_call,
  },
  Component: EventRollCall as FunctionComponent<{
    eventId: string;
    isOrganizer: boolean | null | undefined;
  }>,
};
