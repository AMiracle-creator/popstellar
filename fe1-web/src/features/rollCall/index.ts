import { RollCallEventType } from './components';
import * as functions from './functions';
import { RollCallHooks } from './hooks';
import { RollCallConfiguration, RollCallInterface, ROLLCALL_FEATURE_IDENTIFIER } from './interface';
import { configureNetwork } from './network';
import { rollCallReducer } from './reducer';
import * as screens from './screens';

/**
 * Configures the roll call feature
 *
 * @param configuration - The configuration object for the rollcall feature
 */
export function configure(configuration: RollCallConfiguration): RollCallInterface {
  configureNetwork(configuration);

  return {
    identifier: ROLLCALL_FEATURE_IDENTIFIER,
    eventTypes: [RollCallEventType],
    screens,
    functions,
    hooks: {
      useRollCallsByLaoId: RollCallHooks.useRollCallsByLaoId,
    },
    context: {
      useCurrentLaoId: configuration.useCurrentLaoId,
      makeEventByTypeSelector: configuration.makeEventByTypeSelector,
      generateToken: configuration.generateToken,
      hasSeed: configuration.hasSeed,
    },
    reducers: {
      ...rollCallReducer,
    },
  };
}
