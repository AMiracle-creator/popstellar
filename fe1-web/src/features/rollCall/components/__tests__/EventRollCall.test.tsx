import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { combineReducers, createStore } from 'redux';

import { mockNavigate } from '__mocks__/useNavigationMock';
import MockNavigator from '__tests__/components/MockNavigator';
import { mockLaoIdHash, mockLao, mockLaoId } from '__tests__/utils';
import FeatureContext from 'core/contexts/FeatureContext';
import { Hash, Timestamp } from 'core/objects';
import { addEvent, eventsReducer, makeEventByTypeSelector } from 'features/events/reducer';
import { connectToLao, laoReducer } from 'features/lao/reducer';
import { mockRollCall } from 'features/rollCall/__tests__/utils';
import { RollCallReactContext, ROLLCALL_FEATURE_IDENTIFIER } from 'features/rollCall/interface';
import { addRollCall, rollCallReducer, updateRollCall } from 'features/rollCall/reducer';
import { generateToken } from 'features/wallet/objects';
import { getWalletState, walletReducer } from 'features/wallet/reducer';
import STRINGS from 'resources/strings';

import { requestOpenRollCall, requestReopenRollCall } from '../../network';
import { RollCall, RollCallStatus } from '../../objects';
import EventRollCall from '../EventRollCall';

const ID = new Hash('rollCallId');
const NAME = 'myRollCall';
const LOCATION = 'location';
const TIMESTAMP_START = new Timestamp(1620255600);
const TIMESTAMP_END = new Timestamp(1620357600);
const ATTENDEES = ['attendee1', 'attendee2'];

const createStateWithStatus: any = (mockStatus: RollCallStatus) => {
  return {
    id: ID.valueOf(),
    eventType: RollCall.EVENT_TYPE,
    start: TIMESTAMP_START.valueOf(),
    name: NAME,
    location: LOCATION,
    creation: TIMESTAMP_START.valueOf(),
    proposedStart: TIMESTAMP_START.valueOf(),
    proposedEnd: TIMESTAMP_END.valueOf(),
    status: mockStatus,
    attendees: ATTENDEES,
    idAlias: mockStatus === RollCallStatus.CREATED ? undefined : ID.valueOf(),
  };
};

const mockRollCallClosed = RollCall.fromState(createStateWithStatus(RollCallStatus.CLOSED));
const mockRollCallCreated = RollCall.fromState(createStateWithStatus(RollCallStatus.CREATED));
const mockRollCallOpened = RollCall.fromState(createStateWithStatus(RollCallStatus.OPENED));

jest.mock('features/rollCall/network', () => {
  const actualNetwork = jest.requireActual('features/rollCall/network');
  return {
    ...actualNetwork,
    requestOpenRollCall: jest.fn(() => Promise.resolve()),
    requestReopenRollCall: jest.fn(() => Promise.resolve()),
  };
});

// set up mock store
const mockStore = createStore(
  combineReducers({ ...laoReducer, ...eventsReducer, ...rollCallReducer, ...walletReducer }),
);
mockStore.dispatch(connectToLao(mockLao.toState()));
const mockRollCallState = mockRollCallCreated.toState();

mockStore.dispatch(
  addEvent(mockLaoId, {
    eventType: RollCall.EVENT_TYPE,
    id: mockRollCallState.id,
    idAlias: mockRollCallState.idAlias,
    start: mockRollCall.start.valueOf(),
    end: mockRollCall.end.valueOf(),
  }),
);
mockStore.dispatch(addRollCall(mockRollCallState));

const contextValue = {
  [ROLLCALL_FEATURE_IDENTIFIER]: {
    useCurrentLaoId: () => mockLaoIdHash,
    makeEventByTypeSelector,
    generateToken,
    hasSeed: () => getWalletState(mockStore.getState()).seed !== undefined,
  } as RollCallReactContext,
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('EventRollCall', () => {
  it('should correctly render', () => {
    mockStore.dispatch(updateRollCall(mockRollCallCreated.toState()));

    const Screen = () => (
      <EventRollCall
        eventId={mockRollCallCreated.id.valueOf()}
        start={mockRollCallCreated.start.valueOf()}
        end={mockRollCallCreated.end.valueOf()}
        isOrganizer={false}
      />
    );
    const obj = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <MockNavigator component={Screen} />
        </FeatureContext.Provider>
      </Provider>,
    );

    expect(obj.toJSON()).toMatchSnapshot();
  });

  it('should call requestOpenRollCall when the open button is clicked', () => {
    mockStore.dispatch(updateRollCall(mockRollCallCreated.toState()));

    const Screen = () => (
      <EventRollCall
        eventId={mockRollCallCreated.id.valueOf()}
        start={mockRollCallCreated.start.valueOf()}
        end={mockRollCallCreated.end.valueOf()}
        isOrganizer
      />
    );
    const obj = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <MockNavigator component={Screen} />
        </FeatureContext.Provider>
      </Provider>,
    );

    const openRollCallButton = obj.getByText(STRINGS.roll_call_open);
    fireEvent.press(openRollCallButton);
    expect(requestOpenRollCall).toHaveBeenCalledTimes(1);
  });

  it('should call requestReopenRollCall when the reopen button is clicked', () => {
    mockStore.dispatch(updateRollCall(mockRollCallClosed.toState()));

    const Screen = () => (
      <EventRollCall
        eventId={mockRollCallClosed.id.valueOf()}
        start={mockRollCallClosed.start.valueOf()}
        end={mockRollCallClosed.end.valueOf()}
        isOrganizer
      />
    );
    const obj = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <MockNavigator component={Screen} />
        </FeatureContext.Provider>
      </Provider>,
    );

    const reopenRollCallButton = obj.getByText(STRINGS.roll_call_reopen);
    fireEvent.press(reopenRollCallButton);
    expect(requestReopenRollCall).toHaveBeenCalledTimes(1);
  });

  it('should navigate to RollCallOpened when scan attendees button is clicked', () => {
    mockStore.dispatch(updateRollCall(mockRollCallOpened.toState()));

    const Screen = () => (
      <EventRollCall
        eventId={mockRollCallOpened.id.valueOf()}
        isOrganizer
        start={mockRollCallOpened.start.valueOf()}
        end={mockRollCallOpened.end.valueOf()}
      />
    );
    const obj = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <MockNavigator component={Screen} />
        </FeatureContext.Provider>
      </Provider>,
    );

    const scanAttendeesButton = obj.getByText(STRINGS.roll_call_scan_attendees);
    fireEvent.press(scanAttendeesButton);
    expect(mockNavigate).toHaveBeenCalledTimes(1);
  });
});
