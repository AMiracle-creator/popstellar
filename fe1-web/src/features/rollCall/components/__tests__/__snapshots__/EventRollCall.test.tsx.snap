// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventRollCall should correctly render 1`] = `
Array [
  <Text>
    Roll Call
  </Text>,
  <Text>
    Closed
  </Text>,
  <Text>
    Attendees are:
  </Text>,
  <Text>
    a
    t
    t
    e
    n
    d
    e
    e
    1
  </Text>,
  <Text>
    a
    t
    t
    e
    n
    d
    e
    e
    2
  </Text>,
]
`;
